### ISSUE TITLE
Support for relative URLs

### ISSUE THREAD

**Issue by jandrieu** [tobie/pr-preview#144]
Currently, images (and other assets) referenced by relative URLs fail to load properly, making embedded images broken links in the preview.

I believe that if we set the base to point to the PR's raw files, this problem goes away.

```html
<base href="https://raw.githubusercontent.com/[user]/[repo]/[commit]/">
```
Where the user, repo, and commit point to that particular PR (which I got by inspecting the URLs in the files tab of the PR page.

If I manually add this base in the source, it works. Of course, it's crazy making to put details about the PR into the content of the fork itself (you have to create the PR, then update the fork), but it proved that a good <base> element could solve the problem.

It may be that this is ALWAYS the right approach (base SHOULD be the PR), but at a minimum, I'd like a way to turn it on in a config file if it isn't a good default for some reason.

Can anyone point me to the code where this might be handled? Happy to try a PR, but I'm new to the code base.

I note that it also worked as 
```html
<base href="https://raw.githubusercontent.com/[user]/[repo]/[branch]">
```
When using the incoming fork's user and repo. As a hack, that makes it a bit easier for the PR creator as they can just refer to the right branch in their fork, without needing to know about any particular PR. Unfortunately, it would mean that the PR preview is always looking at the current branch, which may not be correct historically. You could view an old, closed PR, and get the wrong view of the asset. Linking to the PR's commit doesn't have that problem.

This seems like it would be a generally useful improvement.

Thoughts?

**Comment by tobie** [comment]
Unfortunately, that would [break all in-page anchors](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base#in-page_anchors), no?

**Comment by jandrieu** [comment]
Ah-hah!

Great catch. I didn't realize that an in-page anchor uses the base in that manner. I would have expected the link to not trigger a query back to the server. Clearly the spec says otherwise and browsers behave according to spec.

However, it works if you include the filename in the base:
```html
<base href="https://raw.githubusercontent.com/[user]/[repo]/[commit]/[filename]">
```
You can see this at this PR: https://github.com/w3c/vc-use-cases/pull/147

So, I think this new base will work for everything EXCEPT where the page authors need to manually set base. So, I think if we automatically set it in cases where there is no base, it could work.

**Comment by jandrieu** [comment]
Btw, we are also using ReSpec data-includes in that file, and they were totally broken in the preview. With the base hack, they work.

**Comment by tobie** [comment]
One way around this would be to make this opt-in.

**Comment by tobie** [comment]
So this could be either handled at the PR Preview level:

- a new opt-in option in the config file
- documentation in the README
- Adding a base element to the generated output (probably just searching for the `head` element and appending a `base` element to its first tag).

or as an option on ReSpec or Bikeshed that you can just pass a value to, much like we already pass the URL of the PR, for example.

**Comment by jandrieu** [comment]
@tobie Anything I can do to help move this forward?

I like the option of adding a config parameter, then injecting the base as appropriate by automatically extracting the PR's URL for that file.

FWIW, passing the URL as a parameter is also a reasonable feature, but would not be much of an improvement over my current hack of manually looking up that URL and injecting the BASE in the file directly. If you have to edit the config file to set the URL, then you have to adjust the URL uniquely for every PR (which is essentially the same amount of work).

If you can send me to the right place in the code to try out this feature, I'll take a look.

**Comment by rjpadula** [comment]
Thanks for the engagement. We dove into the PR-Preview code and believe we found out where we could update that to pull in the additional resources to get them up to AWS. However, it didn’t seem like that made sense for a couple of reasons.

First, we aren’t sure the best way to gather all of the relative resources that might be used in a page. One way could be to take the spec-generator output, search it for tags with src attributes that are image types with relative URLs, CSS files, scripts, etc, and then fetch & upload those images to the S3 bucket using /${owner}/${repo}/pull/{$prNumber}/{resourcePath}. This would address resources loaded via static URLs, but not ones that are downloaded programmatically, such as over AJAX.

Second, it seemed like a potential challenge that all of the PR-Preview assets are stored on a single S3 bucket, presumably owned by you. Not only might that become a data capacity problem–storing all the local resources could be a significant storage cost–it also would subject you, personally, to liability from the content of any of those assets, since you (as Tobie) can’t guarantee any given image is appropriate for distribution.

Third, at the moment we aren’t set up to debug Github Apps or AWS integrations. So even if we wanted to dive into some adjustment based on the current architecture, it would take quite a bit of effort just to verify our changes work.


Finally, we did explore using github actions as an alternative, and we have an approach we believe will work, but that would require replacing the entire respec workflow, which feels like more work than we have budget for at the moment. We’ll keep that as a future option, but respec does so much, replacing it would likely be beyond our resources at this time.

**Comment by tobie** [comment]
Yeah. The original idea of PR preview was to manage basic single page specs long before GitHub actions existed. An acceptable workaround for images was to link to images with absolute URLs. If I was to rebuild it from the ground up today, it would be structured very differently.
